plutus-playground-client/src/AjaxUtils.purs

```diff
@@ -1,20 +1,89 @@
 import Data.Argonaut.Parser (jsonParser)
 import Data.Array (intercalate)
-import Data.Either (Either(..))
-import Data.Generic (class Generic)
+import Data.Either (Either(Right, Left), note)
+import Data.Generic (class Generic, GenericSignature(SigProd), GenericSpine(SProd), fromSpine, isValidSpine, toSignature, toSpine)
+import Data.List as List
+import Data.List.NonEmpty as NonEmpty
+import Data.Maybe (Maybe(..), fromMaybe)
+import Gist (GistId)
 import Halogen.HTML (HTML, br_, div_, pre_, text)
 import Language.Haskell.Interpreter (CompilationError(..))
 import Network.HTTP.StatusCode (StatusCode(..))
-import Prelude (bind, pure, show, ($), (<$>), (<>), (>>>))
+import Playground.API (TokenId)
+import Playground.Server (SPParams_(..))
+import Prelude (bind, pure, show, unit, ($), (<$>), (<>), (>>>))
 import Servant.PureScript.Affjax (AjaxError, ErrorDescription(ConnectionError, DecodingError, ParsingError, UnexpectedHTTPStatus), runAjaxError)
-import Servant.PureScript.Settings (SPSettingsDecodeJson_(..), SPSettingsEncodeJson_(..), SPSettings_(..))
+import Servant.PureScript.Settings (SPSettingsDecodeJson_(..), SPSettingsEncodeJson_(..), SPSettingsToUrlPiece_(..), SPSettings_(..), URLPiece, defaultSettings, gDefaultToURLPiece)
+import Type.Proxy (Proxy(..))
+
+ajaxSettings :: SPSettings_ SPParams_
+ajaxSettings = SPSettings_ $ settings
+  { toURLPiece = SPSettingsToUrlPiece_ gCustomToURLPiece
+  , decodeJson = SPSettingsDecodeJson_ (genericDecodeJson $ Options $ options {userDecoding = userDecoding})
+  }
+  where
+    SPSettings_ settings = defaultSettings $ SPParams_ { baseURL: "/api/" }
+    Options options = Aeson.options
+
+ourDecodeJson :: forall a. Generic a => Json -> Either String a
+ourDecodeJson =
+  let (SPSettings_ {decodeJson: SPSettingsDecodeJson_ decodeJson}) = ajaxSettings
+  in decodeJson
+
+userDecoding :: Options -> GenericSignature -> Json -> Maybe (Either String GenericSpine)
+userDecoding opts sig json =
+  decodeTokenIdLists opts sig json
+  <|>
+  Aeson.userDecoding opts sig json
+
+decodeTokenIdLists :: Options -> GenericSignature -> Json -> Maybe (Either String GenericSpine)
+decodeTokenIdLists opts sig@(SigProd "Data.List.Types.List" [{sigValues: [a, _]}, _]) json =
+  runExceptT do
+    case a unit of
+      (SigProd "Playground.API.TokenId" _) -> do
+        tokenIds :: Array TokenId <- ExceptT $ Just $ ourDecodeJson json
+        pure $ toSpine $ List.fromFoldable tokenIds
+      _ -> empty
+decodeTokenIdLists opts (SigProd "Data.List.Types.NonEmptyList" [{sigValues: [l]}]) json =
+  runExceptT do
+    case l unit of
+      (SigProd "Data.List.Types.List" [{sigValues: [a, _]}, _])  -> do
+        case a unit of
+          (SigProd "Playground.API.TokenId" _) -> do
+            tokenIds :: Array TokenId <- ExceptT $ Just $ ourDecodeJson json
+            nonEmpty <- ExceptT $ Just $ note "List is empty, expecting non-empty" $ NonEmpty.fromList $ List.fromFoldable tokenIds
```

Comment:
can't you use https://pursuit.purescript.org/packages/purescript-lists/4.0.1/docs/Data.List.NonEmpty#v:fromFoldable ?

Reply:
I can, yes - thank you!